### ISSUE TITLE
flag "--address" does not work when using "istioctl dashboard"

### ISSUE THREAD

**Issue by tkna** [istio/istio#26347]
**Bug description**
Flag "--address" does not work when using "istioctl dashboard".

```
# istioctl dashboard kiali --address 0.0.0.0
http://localhost:33197/kiali
```
I expected istioctl to listen on 0.0.0.0, but it is still using localhost.

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**
istioctl binds any ip address specified by flag "--address"

**Steps to reproduce the bug**
```
istioctl dashboard kiali --address 0.0.0.0
```

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
# istioctl version --remote
client version: 1.6.7
control plane version: 1.6.7
data plane version: 1.6.7 (9 proxies)
```
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:48:36Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
```
istioctl install --set profile=demo
```
by following "Getting Started" (https://istio.io/latest/docs/setup/getting-started/)

**Environment where bug was observed (cloud vendor, OS, etc)**
CentOS 7.6 (VirtualBox VM)
minikube v1.10.1 (1 node, vm-driver=none)
Docker Engine - Community    Version:19.03.8

**Comment by chupark** [comment]
So do I. But 'netstat -naopt' display correctly
![localhost](https://user-images.githubusercontent.com/32254293/89868893-7f243480-dbee-11ea-8473-cc5454134a71.png)

**Comment by esnible** [comment]
I am testing on OSX, using the `envoy` dashboard, because I didn't install Kiali.

When I use `--address 127.0.01`, `netstat -an -ptcp | grep LISTEN` shows 
```
tcp4       0      0  127.0.0.1.15000        *.*                    LISTEN
```

When I use `-address 0.0.0.0`, `netstat -an -ptcp | grep LISTEN` shows 
```
tcp4       0      0  *.15000                *.*                    LISTEN 
```

It is true that the address Istio tells the user to visit is always `localhost`.  It can't print the value of the address ... a connection to 0.0.0.0 will fail because that isn't an IPv4 address.  External users should use the address of the machine that ran the `dashboard` command.

Does that answer your question?

**Comment by tkna** [comment]
@chupark @esnible 
It looks like just an issue of display. 
I confirmed that the address and port I specified was listened by istioctl using netstat, too.

@esnible Thank you for your reply. I'm using a specific IPv4 address of the machine (not 0.0.0.0) when I connect to the dashboard.

I think this is caused by line 144 of portforwarder.go. (it should be 'localAddress' instead of 'defaultLocalAddress')
https://github.com/istio/istio/blob/bf4903ee29d256b943b40d2d9c5656230ab8fe4f/pkg/kube/portforwarder.go#L139-L145

**Comment by esnible** [comment]
I think I understand now.

For IPv4, we should show and open `localhost` for `--address 0.0.0.0`, but we MUST show and open 192.168.1.1 for `--address 192.168.1.1`.  I can handle that.

For IPv6 I don't know what to do.

**Comment by tkna** [comment]
@esnible 
> For IPv4, we should show and open localhost for --address 0.0.0.0,

Is there any security reason for that?
It sounds natural to me that we show and open 0.0.0.0 or * (=not only localhost, but also all the ip addresses on the node, like *.15000 in your netstat output) when `--address 0.0.0.0` is supplied.
(Sorry if I don't understand what you mean)

**Comment by esnible** [comment]
> It sounds natural to me that we show and open 0.0.0.0 or * (=not only localhost, but also all the ip addresses on the node, like *.15000 in your netstat output) when `--address 0.0.0.0` is supplied.
> (Sorry if I don't understand what you mean)

What I mean by "open" is that after the port is listening, _istioctl_ starts the opens a browser window for a URL.  The OS command to create/open a browser window is, depending on the OS, `xdg-open`, `rundll32`, or `open`.  We can't create a browser window for `http://0.0.0.0`.  That's a listening IP address, not a contactable one.

I am open to suggestions for this issue.  I believe the use case is where the user is logged into a remote machine without a browser, and the user does `istioctl dashboard`, and the output message gives a `localhost` URL that won't work on the user's laptop?  This doesn't affect the case where the user opens a browser on the `istioctl` machine.

When I am in this situation I run `ifconfig` on the machine running istioctl and open my local browser to the address provided by ifconfig, the port provided by istioctl.  It is hard for istioctl to figure out the address of the machine it is running on.  VMs these days have multiple addresses, and they are on VPNs.  Can anyone suggest a way to improve istioctl's output when `--address 0.0.0.0` is being used?

**Comment by tkna** [comment]
> What I mean by "open" is that after the port is listening, istioctl starts the opens a browser window for a URL.

Now I understand. In my case, istioctl machine is a Virtualbox VM with no browser on my laptop, so I try to connect the dashboard using my local browser after I open up the ports by istioctl. Of course, in this case

1) there's no need to use `--address 0.0.0.0` because I know another specific IP address (e.g. 192.168.X.X) on the VM. Instead, I can just use `--address 192.168.X.X`.
2) there's no need to "open a browser window" by istioctl.

as you mentioned.

**Comment by esnible** [comment]
@tkna Shall I close this issue?

**Comment by tkna** [comment]
Actually, I have created a new version where istioctl's output is fixed at least, but it's OK to close since it's a minor issue.

https://github.com/istio/istio/compare/master...tkna:fix_istioctl_dashboard_output

**Comment by esnible** [comment]
@tkna I looked at that link.  It will do better than current istioctl with `--address 9.1.2.3` but worse with `--address 0.0.0.0`.  I will close this, but if anyone wants to work on this let me know.

In the UX meeting yesterday we discussed that some of the dashboard commands might be going away, because they reflect demo add-ons and the current implementation is not suitable for production.  We may revisit the output of these commands then.
